pkg/controller/webhookbootstrap/controller.go

```diff
@@ -0,0 +1,488 @@
+	if !c.certificateRequiresIssuance(ctx, log, secret, pk, crt) {
+		return nil
+	}
+
+	signedCert, err := c.selfSignCertificate(crt, pk)
+	if err != nil {
+		log.Error(err, "error signing certificate")
+		return err
+	}
+
+	return c.updateSecret(secret, pkData, signedCert, signedCert)
+}
+
+func (c *controller) syncServingSecret(ctx context.Context, secret *corev1.Secret) error {
+	log := logf.FromContext(ctx, "ca-secret")
+	log = logf.WithResource(log, secret)
+	crt := buildServingCertificate(secret, c.webhookDNSNames)
+
+	// first fetch the CA private key & certificate
+	caSecret, err := c.secretLister.Secrets(c.webhookNamespace).Get(c.webhookCASecret)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Error(err, "ca secret does not yet exist")
+			// TODO: automatically sync the serving secret when the ca secret
+			//       is updated and return nil here instead
+			return err
+		}
+		return err
+	}
+
+	caPKData := readSecretDataKey(caSecret, corev1.TLSPrivateKeyKey)
+	caPK, err := pki.DecodePrivateKeyBytes(caPKData)
+	if err != nil {
+		log.Error(err, "Error decoding CA private key")
+		return err
+	}
+
+	caCertData := readSecretDataKey(caSecret, corev1.TLSCertKey)
+	caCert, err := pki.DecodeX509CertificateBytes(caCertData)
+	if err != nil {
+		log.Error(err, "Error decoding CA certificate data")
+		return err
+	}
+
+	// read the existing private key
+	pkData := readSecretDataKey(secret, corev1.TLSPrivateKeyKey)
+	if pkData == nil {
+		log.Info("Generating new private key")
+		return c.generatePrivateKey(crt, secret)
+	}
+	pk, err := pki.DecodePrivateKeyBytes(pkData)
+	if err != nil {
+		log.Info("Regenerating new private key")
+		return c.generatePrivateKey(crt, secret)
+	}
+	// read the existing certificate
+	if !c.certificateRequiresIssuance(ctx, log, secret, pk, crt) {
+		log.Info("Serving certificate already up to date")
+		return nil
+	}
+
+	// TODO: check to make sure the serving certificate is signed by the CA
+	certData, err := c.signCertificate(crt, pk, caPK, caCert)
+	if err != nil {
+		log.Error(err, "error signing certificate")
+		return err
+	}
+
+	return c.updateSecret(secret, pkData, caCertData, certData)
+}
+
+func (c *controller) certificateRequiresIssuance(ctx context.Context, log logr.Logger, secret *corev1.Secret, pk crypto.Signer, crt *cmapi.Certificate) bool {
+	// read the existing certificate
+	crtData := readSecretDataKey(secret, corev1.TLSCertKey)
+	if crtData == nil {
+		log.Info("Issuing webhook certificate")
+		return true
+	}
+	cert, err := pki.DecodeX509CertificateBytes(crtData)
+	if err != nil {
+		log.Info("Re-issuing webhook certificate")
+		return true
+	}
+
+	// ensure private key is valid for certificate
+	matches, err := pki.PublicKeyMatchesCertificate(pk.Public(), cert)
+	if err != nil {
+		log.Error(err, "internal error checking certificate, re-issuing certificate")
+		return true
+	}
+	if !matches {
+		log.Info("certificate does not match private key, re-issuing")
```

Comment:
nit: Not too important, just there is a mix of upper and lower cases as start of log messages